field/ordered_map.go

```diff
@@ -8,15 +8,15 @@ import (
 )
 
 // Custom type to sort keys in resulting JSON
-type OrderedMap map[int]Field
+type OrderedMap map[string]Field
```

Comment:
@krishishah with the change, we don't get proper order in JSON for numeric IDs. Here is an example of what we may get (when we have 17 and 107 as IDs):

```json
{
  "0":"0100",
  "1":"D00080000000000000000000002000000000000000000000",
  "107":"102",
  "17":"101",
  "2":"4242424242424242",
  "4":"100"
}
```

I think that we may keep OrderedMap as is for numeric IDs and just use default JSON marshalling for a composite field without sorting. Or, we may add `OrderedStringMap` in addition to `Ordered(Int)Map`. What do you think?

Reply:
Great spot @alovak !

I don't think we should should maintain two different implementations of `OrderedMap` (i.e. `OrderedIntMap` and `OrderedStringMap`).

Ideally, we'd just generify `OrderedMap` to accommodate for different sorting requirements.

Thankfully, the `sort` pkg gives us much of the behaviour we need.

The way I see it is we have two options:

1. We can change the `OrderedMap` type to:

```Go
// Custom type to sort keys in resulting JSON                                                                                                                                                           
type OrderedMap struct {                                                                                                                                                                                
    Fields map[string]Field                                                                                                                                                                             
    Sort sort.StringSlice // "github.com/moov-io/iso8583/sort"                                                                                                                                                                          
}
```

The big change here would be to call `om.Sort(keys)` instead of `strings.Sort(keys)`. Our two big users of this type are `Message` and `field.Composite`. Each of them can quite easily pass in the correct `sort` funcs to determine the ordering of those keys.

One optimisation we can make is to push the sorting of these keys onto the callers of `OrderedMap`. This is more efficient to do for `fields.Composite` because we already maintain an ordered list of fields and this leads us onto the next option.

2. Instead of holding a `map[string]Field`, `OrderedMap` could now become:

```go
type KeyVal struct {
    Key string
    Val Field \\ this could even be an interface{} if we want to generify it further.
}

// Define an ordered map
type OrderedMap []KeyVal
```

Users of this field can populate `[]KeyVal` in an order of their choice and delegate marshalling to `OrderedMap.MarshalJSON()`

What are your thoughts?